test/parallel/test-fs-rm.js

```diff
@@ -287,74 +286,154 @@ function removeAsync(dir) {
   // IBMi has a different access permission mechanism
   // This test should not be run as `root`
   if (!common.isIBMi && (common.isWindows || process.getuid() !== 0)) {
-    function makeDirectoryReadOnly(dir, mode) {
-      let accessErrorCode = 'EACCES';
-      if (common.isWindows) {
-        accessErrorCode = 'EPERM';
-        execSync(`icacls ${dir} /deny "everyone:(OI)(CI)(DE,DC)"`);
-      } else {
-        fs.chmodSync(dir, mode);
+    // Check that deleting a file that cannot be accessed using rmsync throws:
+    // https://github.com/nodejs/node/issues/38683
+    {
+      function isValidState(exists, err) {
+        // On Windows, we are allowed to access and modify the contents of a
+        // read-only folder.
+        return common.isWindows ?
+          (exists === false && err === null) :
+          (exists === true && err?.code === 'EACCES');
       }
-      return accessErrorCode;
-    }
 
-    function makeDirectoryWritable(dir) {
-      if (fs.existsSync(dir)) {
-        if (common.isWindows) {
-          execSync(`icacls ${dir} /remove:d "everyone"`);
-        } else {
-          fs.chmodSync(dir, 0o777);
+      {
+        const dirname = nextDirPath();
+        const filePath = path.join(dirname, 'text.txt');
+
+        fs.mkdirSync(dirname, { recursive: true });
+        fs.writeFileSync(filePath, 'hello');
+
+        fs.chmodSync(filePath, 0o444);
+        fs.chmodSync(dirname, 0o444);
+
+        let err = null;
+
+        try {
+          fs.rmSync(filePath, { force: true });
+        } catch (_err) {
+          err = _err;
+        }
+
+        try {
+          fs.chmodSync(dirname, 0o777);
+        } catch {
+        }
+
+        try {
+          fs.chmodSync(filePath, 0o777);
+        } catch {
+        }
+
+        if (!isValidState(fs.existsSync(filePath), err)) {
+          throw err;
         }
       }
-    }
 
-    {
-      // Check that deleting a file that cannot be accessed using rmsync throws
-      // https://github.com/nodejs/node/issues/38683
-      const dirname = nextDirPath();
-      const filePath = path.join(dirname, 'text.txt');
-      try {
+      {
+        const dirname = nextDirPath();
+        const filePath = path.join(dirname, 'text.txt');
+
         fs.mkdirSync(dirname, { recursive: true });
         fs.writeFileSync(filePath, 'hello');
-        const code = makeDirectoryReadOnly(dirname, 0o444);
-        assert.throws(() => {
-          fs.rmSync(filePath, { force: true });
-        }, {
-          code,
-          name: 'Error',
-        });
-      } finally {
-        makeDirectoryWritable(dirname);
+
+        fs.chmodSync(filePath, 0o444);
+        fs.chmodSync(dirname, 0o444);
+
+        fs.rm(filePath, { force: true }, common.mustCall((err) => {
+          try {
+            fs.chmodSync(dirname, 0o777);
+          } catch {
+          }
+
+          try {
+            fs.chmodSync(filePath, 0o777);
+          } catch {
+          }
```

Comment:
nit: keeping those in one line to make it clearer that those revert the action of lines 340 and 341
```suggestion
        try { fs.chmodSync(dirname, 0o777); } catch {}
        try { fs.chmodSync(filePath, 0o777); } catch {}
```